q2cli/cache.py

```diff
@@ -0,0 +1,281 @@
+            self._cache_current_state(current_requirements)
+            with open(state_path, 'r') as fh:
+                return json.load(fh)
+
+    # NOTE: The private methods below are all used internally within
+    # `_get_cached_state`.
+
+    def _get_current_requirements(self):
+        """Includes installed versions of q2cli and QIIME 2 plugins."""
+        import os
+        import pkg_resources
+        import q2cli
+
+        reqs = {
+            pkg_resources.Requirement.parse('q2cli == %s' % q2cli.__version__)
+        }
+
+        # A distribution (i.e. Python package) can have multiple plugins, where
+        # each plugin is its own entry point. A distribution's `Requirement` is
+        # hashable, and the `set` is used to exclude duplicates. Thus, we only
+        # gather the set of requirements for all installed Python packages
+        # containing one or more plugins. It is not necessary to track
+        # individual plugin names and versions in order to determine if the
+        # cache is outdated.
+        #
+        # TODO: this code is copied from
+        # `qiime.sdk.PluginManager.iter_entry_points`. Importing QIIME is
+        # currently slow, and it adds ~600-700ms to any CLI command. This makes
+        # the CLI pretty unresponsive, especially when running help/informative
+        # commands. Uncomment the following lines when
+        # https://github.com/qiime2/qiime2/issues/151 is fixed:
+        #
+        # for ep in qiime.sdk.PluginManager.iter_entry_points():
+        #     reqs.add(ep.dist.as_requirement())
+        #
+        for entry_point in pkg_resources.iter_entry_points(
+                group='qiime.plugins'):
+            if entry_point.name != 'dummy-plugin' or 'QIIMETEST' in os.environ:
+                reqs.add(entry_point.dist.as_requirement())
+
+        return reqs
+
+    def _get_cached_requirements(self):
+        import os.path
+        import pkg_resources
+
+        path = os.path.join(self._cache_dir, 'requirements.txt')
+
+        if not os.path.exists(path):
+            # No cached requirements. The empty set will always trigger a cache
+            # refresh because the current requirements will, at minimum,
+            # contain q2cli.
+            return set()
+        else:
+            with open(path, 'r') as fh:
+                contents = fh.read()
+            try:
+                return set(pkg_resources.parse_requirements(contents))
+            except pkg_resources.RequirementParseError:
+                # Unreadable cached requirements, trigger a cache refresh.
+                return set()
+
+    def _cache_current_state(self, requirements):
+        import json
+        import os.path
+        import click
+
+        click.secho(
+            "QIIME is caching your current deployment for improved "
+            "performance. This may take a few moments and should only happen "
+            "once per deployment.", fg='yellow')
```

Comment:
Should this be `stderr`? I'm honestly not sure, but it strikes me as something you could filter out if you were inspecting an `info` call.

It's a shame there isn't a `stdinfo` fd, that would be perfect for these kinds of things.